python/cuml/cuml/tests/conftest.py

```diff
@@ -234,16 +235,24 @@ def housing_dataset():
     return X, y, feature_names
 
 
-@pytest.fixture(scope="module")
+@functools.cache
+def get_boston_data():
+    return pd.read_csv(
+        "https://raw.githubusercontent.com/scikit-learn/scikit-learn/baf828ca126bcb2c0ad813226963621cafe38adb/sklearn/datasets/data/boston_house_prices.csv",
+        header=None,
+    )  # noqa: E501
+
+
+@pytest.fixture(scope="session")
 def deprecated_boston_dataset():
     # dataset was removed in Scikit-learn 1.2, we should change it for a
     # better dataset for tests, see
     # https://github.com/rapidsai/cuml/issues/5158
 
-    df = pd.read_csv(
-        "https://raw.githubusercontent.com/scikit-learn/scikit-learn/baf828ca126bcb2c0ad813226963621cafe38adb/sklearn/datasets/data/boston_house_prices.csv",
-        header=None,
-    )  # noqa: E501
+    try:
+        df = get_boston_data()
+    except:  # noqa E722
+        pytest.xfail(reason="Error fetching Boston housing dataset")
```

Comment:
Is there something else we can do here? My feeling is that by `xfail`ing the test we are kinda saying "this will be a test that runs sometimes, but often it won't run. Eventually the test will break but no one will notice because it hardly ever runs. Until one day long after the change that broke the test when it will run and fail. Now someone from the future will have to scratch their head about what the heck just happened to them".

I am not sure what a good solution would be. Here some ideas:
* we now cache the dataset so should run into rate limits less often, so just deal with the occasional failure
* delete all the tests that use boston
* switch all the tests that use boston to some newer dataset (like california?)
* host the boston CSV somewhere we can more reliably fetch it from

Reply:
I chose the first (deal with the occasional failure) because that strategy is used elsewhere in the same tests for robustness. However, the other tests will retry network failures three times. I could do that here too if desired but I didn’t want to make the code more complex.

If the team desires something more substantive like a change in test data or hosting, I would recommend adopting that in a larger-scale rewrite of these tests. We must increase the robustness of our CI, even if it comes at a cost of coverage (if a download fails, just move on). It has been over 130 days since nightly CI passed.

Reply:
Can you remove the `pytest.xfail` then?

Reply:
I think "just deal with the occasional failure" should come at a cost to coverage, and not at a cost to CI pass/fail. CI robustness (signal/noise) is too low to be useful at present, and we need to have _occasional_ passes for nightly CI to be of any value.

There is already an open issue to move away from this long-deprecated dataset, and I would recommend that if the test is xfailing too consistently. In the meantime I can match the same behavior used for other datasets: retry the download 3 times, cache the file locally, and xfail the test if the download does not succeed.

Reply:
I added retry logic in 6ed322a. I now expect this test to be equally robust as the native sklearn `fetch_*` methods.